### DISCUSSION TITLE
Is this possible to predict a delay between sending a command and robot making an action?

### DISCUSSION THREAD

**Discussion by fillif3** [iRobotEducation/create3_docs#569] [Questions]
### How are you connecting to your Create 3?

Wi-Fi (ROS 2)

### Computer(s) Model(s) and Operating System(s)

Asus Vivobook - Ubuntu 22.04

### Which version of ROS 2 is installed on your computer?

Humble

### Which firmware version is installed on your robot?

H.2.5

### Which RMW is your robot running?

CycloneDDS

### Does your robot have an assigned namespace? If so, please share.

No

### Is the robot connected to a network? If so, what is the network type?

Normal WiFi in my office

### Are there multiple Create 3 robots connected to your network?

No

### Is multicast enabled?

No

### What is the Adapter Board's USB/BLE Toggle currently switched to?

None

### Describe your question.

I work with irobot create3 and I noticed that there is a delay between sending a massage and performing an action. I tried to estimate it and the delay is usually less than 2.5 seconds (which is still long) but it can be as much as 8 seconds.

In my work, I try to use the robot to avoid obstacles as close as possible with Model Predictive Control, but it is very difficult to model the delay without being able to predict it.

**Comment by alsora** [comment]
The delay is caused by the network communication.
I don't think you'll be able to predict it reliably.

There are ways to reduce it through configurations of the dds stack.

**Comment by slowrunner** [comment]
> performing an action

Are you using the Create3 actions?  
I avoid for time critical motions.

Have you characterized the /cmd_vel topic delay?  (It appears to be always sub-second unless the Create3 is having one of its 10 second overload crisies)

**Comment by fillif3** [comment]
"The delay is caused by the network communication."
I have a turtlebot 3 in the same network and the delay is lower.

"Are you using the Create3 actions?"
I only use "/cmd_vel" topic to send commands. 

"It appears to be always sub-second"
It is true most of the time but while it is not true then the delay is higher. I just would like to know if it is possible to predict it.

**Reply by alsora** [reply]
> "The delay is caused by the network communication."
I have a turtlebot 3 in the same network and the delay is lower.

The raspberry pi (on the turtlebot) has a more performant wifi module than the create 3.
Using the USB interface rather than the wifi will immediately give lower latency, but ROS doesn't guarantee a deterministic latency.
